### PR TITLE
Add support for CgroupsV2 in beats, migrate away from gosigar

### PR DESCRIPTION
**Pull request by fearful-symmetry**
## What does this PR do?

This PR accomplishes a few different things:

- Adds support for cgroupsV2
- Continues our deprecation of elastic/gosigar, by moving the cgroups code into `libbeat`. A considerable portion of the code here is pre-existing code from `gosigar` that's been reorganized. 
- Moves the `opt` library from `metricbeat/internal/metrics` to `libbeat`, since we use it here too.
- Refactors the existing cgroups V1 code to be more in line with the refactor of the system module, removing the `MapStr` manipulation code in favor of having the data format hard-coded into the structs, and makes the cgroups code overall more "metrics first"

Note that supporting cgroups V1 and V2 involves some compromises, as the two versions structure themselves and report data fairly differently, and as such, most of the metrics APIs require the consumer to differentiate between V1 and V2 if they want to access raw metrics. V1 and V2 can also cooexist on the same system, so this must happen on a pid-by-pid basis.

Also, I'm still testing this on V1 and mixed v1/v2 systems, but the code is otherwise ready for review.

There's also the issue of dashboards. The system Docker dashboard relies on many of the V1 fields that aren't present in V2, and I'm not really sure how to deal with them. Last I tried, we don't really have any mechanism for dashboards to operate with the logic of "use this field if present, otherwise use this other field." We may also just want to re-write the dashboard entirely to just use fields that are present in V1 and V2, at the expense of losing some of the visualizations. 

Also keep in mind that most of `system/process` is going to be aggressively refactored after this, so any issues, unless they're serious, with the code in `libbeat/metric/system/process` and `metricbeat/system/process` will almost certainly be fixed as part of 7.16.

## What's up with cgroupsV2

Cgroups V2 introduces a few major changes compared to cgroups V1, which necessitated a lot of extra code:

- A new hierarchy for processes. This "unified" hierarchy requires new logic for code that wants to track processes in a cgroup
- New controllers. The various resource controllers (`cpuacct`, `memory`, etc) have changed dramatically, and in a few cases have merged with other controllers for V2. This requires entirely new metrics code.

## File structure

For ease of browsing through this PR (Github honestly isn't great at presenting large PRs), here's the breakdown of what's in here:
```
libbeat/metric/system/cgroup // The entirety of the cgroup implementation. The files in this base directory are (mostly) unchanged files from gosigar.
├── cgcommon // Files that are shared by cgv1 and cgv2. Mostly struct definitions and helper code
├── cgv1 // Metrics  for cgroups v1. This is almost entirely code from elastic/gosigar that's been refactored.
├── cgv2 // metrics for cgroups v2. This is entirely new code.
└── testdata // data used by by the various *_test.go files

```
This also includes:

- `libbeat/opt` Which is mostly existing code that was moved from  `metricbeat/internal`.
- Changes to `cmd/instance`, `add_process_metadata` and `add_docker_metadata` to migrate away from `gosigar`
- changes to `metricbeat/internal` to deal with moving the `libbeat/opt` code.

## Why is it important?

Cgroups V2 is coming, and is already default on Fedora. It's supported by most other software at this point.

## Checklist
- [X] My code follows the style guidelines of this project
- [X] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have made corresponding change to the default configuration files
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

- Pull down and build
- Note what cgroups version your host is using. You can see with `grep cgroup /proc/self/mountinfo`
- Test the `system/process` metricset, as well as `add_process_metadata` and `add_docker_metadata`

### CODE REVIEW

File: libbeat/metric/system/process/process.go [hunk 1]
```diff
@@ -369,9 +368,11 @@ func (procStats *Stats) getProcessEvent(process *Process) common.MapStr {
 	}
 
 	if procStats.EnableCgroups {
-		if statsMap := cgroupStatsToMap(process); statsMap != nil {
-			proc["cgroup"] = statsMap
+		statsMap, err := process.RawStats.Format()
+		if err != nil {
+			procStats.logger.Warnf("Getting memory details: %v", err)
 		}
+		proc["cgroup"] = statsMap
```
Comment (jsoriano) [inline libbeat/metric/system/process/process.go:381]: Should we use this statsMap if there was an error?
```suggestion
			procStats.logger.Warnf("Getting memory details: %v", err)
		} else {
		        proc["cgroup"] = statsMap
		}
```

File: libbeat/metric/system/process/process.go [hunk 2]
```diff
@@ -576,19 +531,22 @@ func (procStats *Stats) getSingleProcess(pid int, newProcs ProcsMap) *Process {
 
 	err = process.getDetails(procStats.isWhitelistedEnvVar)
 	if err != nil {
-		procStats.logger.Debugf("Error getting details for process %s with pid=%d: %v", process.Name, process.Pid, err)
+		procStats.logger.Errorf("Error getting details for process %s with pid=%d: %v", process.Name, process.Pid, err)
```
Comment (jsoriano) [inline libbeat/metric/system/process/process.go]: Maybe this was intentionally logged at the debug level? :grimacing: I can imagine this flooding logs in hosts with many short-lived processes.

File: libbeat/metric/system/process/process.go [hunk 3]
```diff
@@ -576,19 +531,22 @@ func (procStats *Stats) getSingleProcess(pid int, newProcs ProcsMap) *Process {
...
 		return nil
 	}
 
 	if procStats.EnableCgroups {
-		cgStats, err := procStats.cgroups.GetStatsForProcess(pid)
+		cgStats, err := procStats.cgroups.GetStatsForPid(pid)
 		if err != nil {
-			procStats.logger.Debug("Error fetching cgroup data for process %s with pid=%d: %v", process.Name, process.Pid, err)
+			procStats.logger.Errorf("Error fetching cgroup data for process %s with pid=%d: %v", process.Name, process.Pid, err)
```
Comment (jsoriano) [inline libbeat/metric/system/process/process.go]: Same here.

File: libbeat/metric/system/process/process.go (hunk 2 above)
Comment (jsoriano) [reply libbeat/metric/system/process/process.go]: Btw, this reminds me to https://github.com/elastic/beats/issues/9111.

File: libbeat/processors/add_docker_metadata/add_docker_metadata.go [hunk 4]
```diff
@@ -22,13 +22,13 @@ package add_docker_metadata
...
 	"os"
-	"path/filepath"
+	"regexp"
 	"strings"
 	"time"
 
 	"github.com/pkg/errors"
 
-	"github.com/elastic/gosigar/cgroup"
+	"github.com/elastic/beats/v7/libbeat/metric/system/cgroup"
```
Comment (jsoriano) [inline libbeat/processors/add_docker_metadata/add_docker_metadata.go:32]: Nit. Move this import to the group of imports below.
```suggestion
	"github.com/elastic/beats/v7/libbeat/metric/system/cgroup"
```

File: libbeat/processors/add_docker_metadata/add_docker_metadata_test.go [hunk 5]
```diff
@@ -33,19 +33,20 @@ import (
...
 )
 
 func init() {
 	// Stub out the procfs.
-	processCgroupPaths = func(_ string, pid int) (map[string]string, error) {
+	processCgroupPaths = func(_ string, pid int) (map[string]cgroup.ControllerPath, error) {
 		switch pid {
 		case 1000:
-			return map[string]string{
-				"cpu": "/docker/FABADA",
```
Comment (jsoriano) [inline libbeat/processors/add_docker_metadata/add_docker_metadata_test.go:44]: https://en.wikipedia.org/wiki/Fabada_asturiana :joy:

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 6]
```diff
@@ -106,25 +107,35 @@ func (p gosigarCidProvider) getProcessCgroups(pid int) (map[string]string, error
...
 			if rs != nil {
 				return rs[1]
 			}
 		}
 	} else {
-		// use string prefix to find cid
+		// else, fall back to a hardcoded regex
+		// In an attempt to not break the user-facing config interface for this processor,
+		// fall back to the config'ed prefixes if we have cgv1, otherwise use regex
+		// This should work with k8s on cgroupsV2, as we're still trying to extract the same container ID
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go:124]: +1, but it'd be good to confirm that.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 7]
```diff
@@ -106,25 +107,35 @@ func (p gosigarCidProvider) getProcessCgroups(pid int) (map[string]string, error
...
 		for _, path := range cgroups {
-			for _, prefix := range p.cgroupPrefixes {
-				if strings.HasPrefix(path, prefix) {
-					return filepath.Base(path)
+			if path.IsV2 {
+				re := regexp.MustCompile(`[\w]{64}`)
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go:127]: Nit. Move this to a package variable so the regexp doesn't need to be compiled every time.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 8]
```diff
@@ -106,25 +107,35 @@ func (p gosigarCidProvider) getProcessCgroups(pid int) (map[string]string, error
...
+				rs := re.FindStringSubmatch(path.ControllerPath)
+				if rs != nil {
+					return rs[0]
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go:130]: Nit. I'd feel safer checking the length.
```suggestion
				if len(rs) > 0 {
					return rs[0]
```

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 9]
```diff
@@ -106,25 +107,35 @@ func (p gosigarCidProvider) getProcessCgroups(pid int) (map[string]string, error
...
+			rs := re.FindStringSubmatch(path.ControllerPath)
 			if rs != nil {
 				return rs[1]
 			}
 		}
 	} else {
-		// use string prefix to find cid
+		// else, fall back to a hardcoded regex
+		// In an attempt to not break the user-facing config interface for this processor,
+		// fall back to the config'ed prefixes if we have cgv1, otherwise use regex
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go:123]: I see `p.cgroupPrefixes` are not used if this is using cgroups v2. This may be a bit unexpected, or this option doesn't make sense under cgroup v2?

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 10]
```diff
@@ -106,25 +107,35 @@ func (p gosigarCidProvider) getProcessCgroups(pid int) (map[string]string, error
 // ID is found then an empty string is returned.
 // Example:
 // /kubepods/besteffort/pod9b9e44c2-00fd-11ea-95e9-080027421ddf/2bb9fd4de339e5d4f094e78bb87636004acfe53f5668104addc761fe4a93588e
-func (p gosigarCidProvider) getCid(cgroups map[string]string) string {
+func (p gosigarCidProvider) getCid(cgroups map[string]cgroup.ControllerPath) string {
 	// if regex defined use it to find cid
 	if len(p.cgroupRegex) != 0 {
 		re := regexp.MustCompile(p.cgroupRegex)
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go:113]: No need to fix this in this PR, but if `p.cgroupRegex` comes from user configuration we shouldn't use `regexp.MustCompile`, this will panic if the user configures an invalid regex.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go [hunk 11]
```diff
@@ -106,25 +107,35 @@ func (p gosigarCidProvider) getProcessCgroups(pid int) (map[string]string, error
...
-		// use string prefix to find cid
+		// else, fall back to a hardcoded regex
+		// In an attempt to not break the user-facing config interface for this processor,
+		// fall back to the config'ed prefixes if we have cgv1, otherwise use regex
+		// This should work with k8s on cgroupsV2, as we're still trying to extract the same container ID
 		for _, path := range cgroups {
-			for _, prefix := range p.cgroupPrefixes {
-				if strings.HasPrefix(path, prefix) {
-					return filepath.Base(path)
+			if path.IsV2 {
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/gosigar_cid_provider.go:126]: Is there any test case covering v2 here?

File: metricbeat/module/system/process/_meta/data.json [hunk 12]
```diff
@@ -11,221 +11,319 @@
...
+        "ppid": 201795,
+        "state": "sleeping",
+        "working_directory": "/usr/share/elastic-agent/data/install/metricbeat-7.15.0-SNAPSHOT-linux-x86_64"
     },
     "service": {
         "type": "system"
     },
     "system": {
         "process": {
             "cgroup": {
```
Comment (jsoriano) [inline metricbeat/module/system/process/_meta/data.json:69]: Is there any field indicating the cgroups version?

File: metricbeat/module/system/process/_meta/fields.yml [hunk 13]
```diff
@@ -398,6 +547,13 @@
                 The number of times that the memory limit (kmem_tcp.limit.bytes)
                 was reached.
 
+            - name: stats.*
+              type: nested
```
Comment (jsoriano) [inline metricbeat/module/system/process/_meta/fields.yml:551]: Are you sure you want to use [nested](https://www.elastic.co/guide/en/elasticsearch/reference/current/nested.html) here? `object` is probably fine, same for the rest of nested fields defined here.
```suggestion
              type: object
              object_type: float
```

**Review by jsoriano — COMMENTED** [review]
Looks quite good. I haven't reviewed the code moved from gosigar, if we want to make any refactor there I would leave it for a separate review. If there are relevant changes there please point to them :slightly_smiling_face:

File: libbeat/cmd/instance/metrics/metrics.go [hunk 14]
```diff
@@ -298,63 +298,107 @@ func reportBeatCgroups(_ monitoring.Mode, V monitoring.Visitor) {
...
+	if cgv == cgroup.CgroupsV1 {
+		selfStats, err := cgroups.GetV1StatsForProcess(pid)
+		if err != nil {
+			logp.Err("error getting cgroup stats: %v", err)
+			return
+		}
+		// GetStatsForProcess returns a nil selfStats and no error when there's no stats
+		if selfStats == nil {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply libbeat/cmd/instance/metrics/metrics.go:317]: This is an awkward corner case, yes, and the logic to deal with it is in `libbeat/metric/system/cgroup/reader.go:CgroupsVersion` Basically, a few OSes, like ubuntu, will enable cgroups V2, but not actually use it by default, so 90% of the corner case is "check to see which of the two cgroups is actually in use." As far as the other fraction, where the process is somehow using both V1 and V2 controllers (I haven't actually seen this in testing), we default to V2.

File: libbeat/cmd/instance/metrics/metrics.go [hunk 15]
```diff
@@ -298,63 +298,107 @@ func reportBeatCgroups(_ monitoring.Mode, V monitoring.Visitor) {
...
 
-	if cpu := selfStats.CPU; cpu != nil {
-		monitoring.ReportNamespace(V, "cpu", func() {
-			if cpu.ID != "" {
-				monitoring.ReportString(V, "id", cpu.ID)
-			}
-			monitoring.ReportNamespace(V, "cfs", func() {
-				monitoring.ReportNamespace(V, "period", func() {
-					monitoring.ReportInt(V, "us", int64(cpu.CFS.PeriodMicros))
+	if cgv == cgroup.CgroupsV1 {
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply libbeat/cmd/instance/metrics/metrics.go:308]: Done.

File: libbeat/metric/system/process/process.go (hunk 2 above)
Comment (fearful-symmetry) [reply libbeat/metric/system/process/process.go]: Yah, just decided to refactor most of this. Even if this library is going away in 7.16, it's still kind of a mess.

File: libbeat/processors/add_docker_metadata/add_docker_metadata_test.go (hunk 5 above)
Comment (fearful-symmetry) [reply libbeat/processors/add_docker_metadata/add_docker_metadata_test.go:44]: Hahahaha

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 6 above)
Comment (fearful-symmetry) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:124]: Yah, I've been struggling to find a prod-like k8s environment that uses V2 for me to test on.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 9 above)
Comment (fearful-symmetry) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:123]: The prefixes are used for different controllers within the cgroup. On V2, the controllers are "unified" so they aren't listed in `/proc/PID/cgroup` We might be able to hack around to make the option do something, but it doesn't really make sense here, I think.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 10 above)
Comment (fearful-symmetry) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:113]: Yah, true...

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 11 above)
Comment (fearful-symmetry) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:126]: I don't think so. We'll probably want to add one...

File: metricbeat/module/system/process/_meta/data.json (hunk 12 above)
Comment (fearful-symmetry) [reply metricbeat/module/system/process/_meta/data.json:69]: Ah! Forgot that.

File: metricbeat/module/system/process/_meta/fields.yml (hunk 13 above)
Comment (fearful-symmetry) [reply metricbeat/module/system/process/_meta/fields.yml:551]: Honestly I spent what felt like half a day trying to make the `fields.yml` happy, and I have no idea if I should use `groups` or `nested` or what. I can change it.

**Comment by fearful-symmetry** [comment]
@jsoriano thanks for the review! addressed most of the issues, and finally tracked down the bug that was breaking CI.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 6 above)
Comment (jsoriano) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:126]: Well, if this is difficult to find maybe this is not so important to cover by now.
I'd suggest to add this specific case to the test plan, @ChrsMark or @MichaelKatsoulis may have more ideas about where a k8s environment with cgroups v2 can be found.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 11 above)
Comment (jsoriano) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:128]: Do you want to add one in this PR, or leave it for a follow up? :slightly_smiling_face:

File: libbeat/metric/system/cgroup/util.go [hunk 16]
```diff
@@ -242,7 +245,23 @@ func (r Reader) ProcessCgroupPaths(pid int) (map[string]ControllerPath, error) {
...
+			// the containerized process won't see the V2 mount
+			// inside /proc/self/mountinfo if docker is using cgroups V1
+			// For this very annoying edge case, revert to the hostfs flag
+			// If it's not set, warn the user that they've hit this.
 			controllerPath := filepath.Join(r.cgroupMountpoints.V2Loc, path)
+			if r.cgroupMountpoints.V2Loc == "" && paths.Paths.Hostfs == "" {
+				logp.L().Errorf(`PID %d contains a cgroups V2 path (%s) but no V2 mountpoint was found.
+This may be because metricbeat is running inside a container on a hybrid system.
+To monitor cgroups V2 processess in this way, mount the unified (V2) hierarchy inside
+the container as /sys/fs/cgroup/unified and start metricbeat with --system.hostfs.`, pid, line)
```
Comment (jsoriano) [inline libbeat/metric/system/cgroup/util.go:260]: Should something about this be added to the docs too?

**Review by jsoriano — APPROVED** [review]
Minor comments added, nothing blocking, but there still seems to be related failures in CI.

File: libbeat/metric/system/process/process.go [hunk 17]
```diff
@@ -497,18 +506,21 @@ func (procStats *Stats) GetOne(pid int) (common.MapStr, error) {
 	}
 
 	newProcs := make(ProcsMap, 1)
-	p := procStats.getSingleProcess(pid, newProcs)
-	if p == nil {
+	p, err := procStats.getSingleProcess(pid, newProcs)
+	if err != nil {
 		return nil, fmt.Errorf("cannot find matching process for pid=%d", pid)
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply libbeat/metric/system/process/process.go:513]: Yah, I'm going to revert the changes here, it's causing issues.

File: libbeat/processors/add_process_metadata/gosigar_cid_provider.go (hunk 11 above)
Comment (fearful-symmetry) [reply libbeat/processors/add_process_metadata/gosigar_cid_provider.go:128]: Ooop, forgot about that. I can add it now.

File: libbeat/metric/system/cgroup/util.go [hunk 18]
```diff
@@ -202,6 +202,10 @@ func SubsystemMountpoints(rootfsMountpoint string, subsystems map[string]struct{
 
 		// V2 option
 		if mount.filesystemType == "cgroup2" {
+			if len(mount.mountpoint) < 3 {
+				// Why is this happening? I don't know.
```
[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply libbeat/metric/system/cgroup/util.go]: Think you're looking at an older version, that's while I was debugging an issue with hybrid cgroups.

File: libbeat/metric/system/cgroup/util.go (hunk 16 above)
Comment (fearful-symmetry) [reply libbeat/metric/system/cgroup/util.go:260]: Yep, good point. I'll add a blurb.

File: libbeat/metric/system/cgroup/util.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (jsoriano) [reply libbeat/metric/system/cgroup/util.go]: Yes, but I was curious about what appeared in `mountinfo` when this happens :slightly_smiling_face:

[the remark this answers is not in the document]
Comment (fearful-symmetry) [reply libbeat/metric/system/cgroup/util.go]: Oh! The cgroups line just...isn't there.

File: libbeat/processors/add_process_metadata/add_process_metadata_test.go [hunk 19]
```diff
@@ -1029,3 +1029,16 @@ func TestPIDToInt(t *testing.T) {
...
+		testMap := map[string]cgroup.ControllerPath{
+			"cpu": {IsV2: true, ControllerPath: "system.slice/docker-2dcbab615aebfa9313feffc5cfdacd381543cfa04c6be3f39ac656e55ef34805.scope"},
+		}
+		return testMap, nil
+	}
+	provider := newCidProvider("", []string{}, "", processCgroupPaths, nil)
+	result, err := provider.GetCid(1)
+	assert.NoError(t, err)
+	assert.Equal(t, "2dcbab615aebfa9313feffc5cfdacd381543cfa04c6be3f39ac656e55ef34805", result)
+}
```
Comment (jsoriano) [inline libbeat/processors/add_process_metadata/add_process_metadata_test.go:1044]: :+1: thanks!

**Review by jsoriano — APPROVED** [review]

File: libbeat/metric/system/cgroup/util.go [hunk 20]
```diff
@@ -0,0 +1,293 @@
...
+			// inside /proc/self/mountinfo if docker is using cgroups V1
+			// For this very annoying edge case, revert to the hostfs flag
+			// If it's not set, warn the user that they've hit this.
+			controllerPath := filepath.Join(r.cgroupMountpoints.V2Loc, path)
+			// Depending on the test environment, Hostfs can either be blank, or `/`
+			if r.cgroupMountpoints.V2Loc == "" && len(paths.Paths.Hostfs) <= 1 {
+				logp.L().Infof(`PID %d contains a cgroups V2 path (%s) but no V2 mountpoint was found.
+This may be because metricbeat is running inside a container on a hybrid system.
+To monitor cgroups V2 processess in this way, mount the unified (V2) hierarchy inside
+the container as /sys/fs/cgroup/unified and start metricbeat with --system.hostfs.`, pid, line)
```
Comment (jsoriano) [inline libbeat/metric/system/cgroup/util.go:260]: @fearful-symmetry in an elastic-agent started with `elastic-package stack` in my laptop, I see continuously (every collection period) this message, I think we should move it to the debug level. In a system when this happens this is going to be continuosly logged.

I use a quite vanilla Ubuntu 18.04.5 LTS, with Linux 4.15, and docker 20.10.7. So maybe this is not so an edge case.

Also, as far as I know, both my host and the docker containers are using cgroups V1 (I don't see any cgroups2 mountpoint or any unified cgroup) so I wonder if this comment about hybrid systems can be misleading in this case.

For reference, this is what I see in logs:
```
2021-08-18T08:15:09.797Z	INFO	cgroup/util.go:257	PID 10 contains a cgroups V2 path (0::/system.slice/containerd.service) but no V2 mountpoint was found.
This may be because metricbeat is running inside a container on a hybrid system.
To monitor cgroups V2 processess in this way, mount the unified (V2) hierarchy inside
the container as /sys/fs/cgroup/unified and start metricbeat with --system.hostfs.
```

This is how `/proc/10/cgroup` looks like in the container:
```
12:freezer:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
11:devices:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
10:cpu,cpuacct:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
9:pids:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
8:hugetlb:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
7:blkio:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
6:perf_event:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
5:cpuset:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
4:memory:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
3:net_cls,net_prio:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
2:rdma:/
1:name=systemd:/docker/b257ef6e9abdac56a8a07bf2e35abc7a064b0638e893bc54ab98d925bba13ecd
0::/system.slice/containerd.service
```

Comment (fearful-symmetry) [reply libbeat/metric/system/cgroup/util.go:260]: @jsoriano 
> `0::/system.slice/containerd.service`

So, that's the V2 controller there. You said you didn't see any unified mountpoint on the system? Can you get me `/proc/self/mountinfo` ?

However, I think you're right that we might want some kind of behavior to prevent this message from getting spammy. Unfortunately, from within a docker container it's really hard to get that info without the user mounting `sysfs` inside the container. Maybe _just_ have it print the message at startup?

Comment (jsoriano) [reply libbeat/metric/system/cgroup/util.go:260]: Oh, you are right, there is some `cgroup2` around, I didn't see it.
```
$ cat /proc/self/mountinfo  | grep cgroup2
34 33 0:29 / /sys/fs/cgroup/unified rw,nosuid,nodev,noexec,relatime shared:10 - cgroup2 cgroup rw,nsdelegate
```

So yeah, let's keep the message but trying to make it less verbose.
